File: sdk/ios_sdk_distribution/Sources/NossoFlutterSDK/NossoFlutterSDK.swift
import Flutter
import SwiftUI
import UIKit

public final class NossoFlutterSDK {
    public static let shared = NossoFlutterSDK()

    private let channelName = "com.empresa.flutter_sdk/auth"
    private lazy var flutterEngine = FlutterEngine(name: "nosso_flutter_sdk_engine")
    private lazy var authChannel = FlutterMethodChannel(
        name: channelName,
        binaryMessenger: flutterEngine.binaryMessenger
    )

    private var isPrepared = false
    private var currentAuthToken: String?

    private init() {}

    public func prepare() {
        guard !isPrepared else {
            return
        }

        flutterEngine.run()
        configureAuthChannel()
        isPrepared = true
    }

    public func start(authToken: String) {
        currentAuthToken = authToken
        prepare()

        authChannel.invokeMethod("authTokenUpdated", arguments: [
            "token": authToken
        ])
    }

    public func makeViewController() -> UIViewController {
        prepare()
        return FlutterViewController(engine: flutterEngine, nibName: nil, bundle: nil)
    }

    private func configureAuthChannel() {
        authChannel.setMethodCallHandler { [weak self] call, result in
            switch call.method {
            case "getAuthToken":
                result(self?.currentAuthToken)
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }
}

public struct NossoFlutterSDKView: UIViewControllerRepresentable {
    private let sdk: NossoFlutterSDK

    public init(sdk: NossoFlutterSDK = .shared) {
        self.sdk = sdk
    }

    public func makeUIViewController(context: Context) -> UIViewController {
        sdk.makeViewController()
    }

    public func updateUIViewController(_ uiViewController: UIViewController, context: Context) {}
}

File: android_sdk_distribution/src/main/kotlin/com/empresa/nossofluttersdk/NossoFlutterSDK.kt
package com.empresa.nossofluttersdk

import android.content.Context
import android.content.Intent
import io.flutter.embedding.android.FlutterActivity
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.FlutterEngineCache
import io.flutter.embedding.engine.dart.DartExecutor
import io.flutter.plugin.common.MethodChannel

class NossoFlutterSDK private constructor(
    context: Context,
) {
    private val appContext = context.applicationContext
    private var currentAuthToken: String? = null

    private val flutterEngine: FlutterEngine by lazy {
        FlutterEngine(appContext).also { engine ->
            engine.dartExecutor.executeDartEntrypoint(
                DartExecutor.DartEntrypoint.createDefault(),
            )
            FlutterEngineCache.getInstance().put(ENGINE_ID, engine)
            configureAuthChannel(engine)
        }
    }

    fun prepare() {
        flutterEngine
    }

    fun start(authToken: String) {
        prepare()
        currentAuthToken = authToken

        MethodChannel(flutterEngine.dartExecutor.binaryMessenger, AUTH_CHANNEL)
            .invokeMethod("authTokenUpdated", mapOf("token" to authToken))
    }

    fun createActivityIntent(context: Context): Intent {
        prepare()
        return FlutterActivity
            .withCachedEngine(ENGINE_ID)
            .build(context)
    }

    private fun configureAuthChannel(engine: FlutterEngine) {
        MethodChannel(engine.dartExecutor.binaryMessenger, AUTH_CHANNEL)
            .setMethodCallHandler { call, result ->
                when (call.method) {
                    "getAuthToken" -> result.success(currentAuthToken)
                    else -> result.notImplemented()
                }
            }
    }

    companion object {
        private const val ENGINE_ID = "nosso_flutter_sdk_engine"
        private const val AUTH_CHANNEL = "com.empresa.flutter_sdk/auth"

        @Volatile
        private var instance: NossoFlutterSDK? = null

        fun initialize(context: Context): NossoFlutterSDK {
            return instance ?: synchronized(this) {
                instance ?: NossoFlutterSDK(context).also { instance = it }
            }
        }

        fun get(): NossoFlutterSDK {
            return requireNotNull(instance) {
                "NossoFlutterSDK.initialize(context) deve ser chamado antes de usar o SDK."
            }
        }
    }
}
